nginx: optimize location blocks (part 2)

Izorkin opened the pull request:
Changes:
 - Remove regexp function from location block. In PR 19438, there was an error in way location blocks was handled.
 - Remove unused location block and add new location block.
 - Move processing of remaining static files to main location block.
 - Renaming `@proxy` block to `@mastodon`.
 - Remove unused header.
 - Remove try_files directive.

Headers for `sw.js.map` file before PR:
```
HTTP/2 200
server: nginx
date: Tue, 01 Nov 2022 16:47:36 GMT
content-type: application/octet-stream
content-length: 297410
etag: "9417r0hndwiqqq6vp5i9ql956bhypr0c"
strict-transport-security: max-age=63072000; includeSubDomains
accept-ranges: bytes
```
After PR:
```
HTTP/2 200
server: nginx
date: Tue, 01 Nov 2022 17:31:44 GMT
content-type: application/octet-stream
content-length: 297410
etag: "9417r0hndwiqqq6vp5i9ql956bhypr0c"
cache-control: public, max-age=604800, must-revalidate
strict-transport-security: max-age=63072000; includeSubDomains
accept-ranges: bytes
```

Izorkin commented:
Fixed location.

Izorkin commented:
cc @Gargron @ClearlyClaire

vivlim commented on dist/nginx.conf:
```diff
@@ -18,8 +18,20 @@ server {
-  location /.well-known/acme-challenge/ { allow all; }
-  location / { return 301 https://$host$request_uri; }
+  # Hide /acme-challenge subdirectory and return 404 on all requests.
+  # It is somewhat more secure than letting Nginx return 403.
```
Can you explain the claim that it is "somewhat more secure" to return a 404 on requests to `/acme-challenge` instead of 403?

I could only find this referenced in a couple of blog posts and the gist you linked, which is six years old.

Izorkin replied:
With a 403 response, the client may assume that the directory /.well-known/acme-challenge/ may be present on the server and used to obtain certificates, but it is not accessible. With a 404 response, it's already harder to know if this directory exists.

Izorkin commented:
cc @Gargron @ClearlyClaire @shleeable

Izorkin commented:
Update nginx configuration.
Now for other static files (500.html, robots.txt, inert.css and etc) located at the site root, the caching parameters are correctly specified.
If there are no static files, the request is now redirected to backend.

Izorkin commented:
cc @Gargron @ClearlyClaire @shleeable @mjankowski

ClearlyClaire commented:
Summing up the changes in this PR:
- changing `/.well-known/acme-challenge/` handling to return 404: as far as I know, there is no security downside to letting other people know that you use acme challenges, or even expose those; I don't think this is worth the change
- renaming `@proxy` to `@mastodon`: I have no strong opinion on this, but I don't think this is worth the change
- dropping `/shortcuts`: that's good, I'm not even sure we ever used this path
- falling back to Rails when a static file has not been found: why? this basically reverts what you did in #19438
- adding cache-control and HSTS to the default path: why not, but what's the rationale?
- moving sw.js handling from its specific case to the default case: why not, but what's the rationale?

Izorkin commented:
> * changing `/.well-known/acme-challenge/` handling to return 404: as far as I know, there is no security downside to letting other people know that you use acme challenges, or even expose those; I don't think this is worth the change

I used this example as a basis:
https://gist.github.com/mingderwang/bdf30ead416c9dc0b29710f864c88517

I thought it would be the best variant.

> * renaming `@proxy` to `@mastodon`: I have no strong opinion on this, but I don't think this is worth the change

If several different web services are used on the server, then there will be less confusion

> * falling back to Rails when a static file has not been found: why? this basically reverts what you did in [nginx: optimize locations #19438](https://github.com/mastodon/mastodon/pull/19438)

I have seen reports of static files no longer opening when using docker containers after a previous change.
I can remove this change.

> * adding cache-control and HSTS to the default path: why not, but what's the rationale?
> * moving sw.js handling from its specific case to the default case: why not, but what's the rationale?

As a result, there will be fewer location blocks, and dynamic content will not be affected because other headers will be triggered there

Izorkin commented:
> * changing `/.well-known/acme-challenge/` handling to return 404: as far as I know, there is no security downside to letting other people know that you use acme challenges, or even expose those; I don't think this is worth the change

> * falling back to Rails when a static file has not been found: why? this basically reverts what you did in [nginx: optimize locations #19438](https://github.com/mastodon/mastodon/pull/19438)

Removed this cahnges.

Updated locations block.

@ClearlyClaire do need to increase the caching time for the `sw.js` file?
Then it will be possible to remove another location block.
In any case, if this file changes on the server, its size and Etag value will change, and the browser will update it.

Izorkin commented:
I misunderstood how the must-revalidate parameter worked, removed caching from the `location /` block.
And added `etag` option.

Izorkin commented:
> I'm not an nginx expert, but wanted to respond to the previous ping I got here ... I don't have a strong opinion, other than that this may be easier to review/merge if you can isolate each change to one commit/PR, rather than multiple motivations/changes all in one?

I don't see the point of breaking pr into parts.

Izorkin commented:
cc @ClearlyClaire @renchap

ThisIsMissEm commented:
I've had a look at this, and whilst I'm familiar with nginx, I'm certainly not an expert, but as it stands this PR is really difficult to review. It looks like this would prevent users from just copying the file across during deployment, as it doesn't do the `try_files` logic by default anymore.

Izorkin committed:
nginx: remove regexp function from location block

In PR 19438, there was an error in way location
blocks was handled.

Izorkin committed:
nginx: remove unused location block

Izorkin committed:
nginx: add new location block

Izorkin committed:
nginx: move processing of remaining static files to main location block

Izorkin committed:
nginx: renaming @proxy block to @mastodon

If multiple sites are hosted on server, a common `@proxy` block
name may cause a conflict if another site uses a block with
same name.

Izorkin committed:
nginx: remove unused header

Headers are not used in the streaming block.

Izorkin commented:
Updated PR and split commits.

Izorkin commented:
cc @ClearlyClaire @renchap

Izorkin commented:
cc @Gargron

Izorkin commented:
cc @ThisIsMissEm @ClearlyClaire

mjankowski commented:
Closing this based on previous feedback of both a) skepticism about changes being worth it, b) request to break up into smaller parts.

I don't have a strong opinion on this, but if you think that some/all of these changes are worth moving forward, can you split up into single PRs per type of change, and note in the PR whether its just a style/rename/format thing, or whether it has priacy/security/performance/etc implications (include benchmarks/links if relevant).

Izorkin commented:
I've already broken the PR down into smaller parts with the necessary comments...
Or haven't you checked?

mjankowski commented:
Can you link to those PRs?

Izorkin commented:
This PR. It's broken down into 10 commits with comments.

mjankowski commented:
Sorry, I should have been more specific -- please break into one per-change-type Pull Request, not just commit on this PR (I assume these 10 commits are doing 3-4 different things?)

Izorkin commented:
Basically, PR is making changes to the block of locations.
I can remove changes related to the block of proxy and upstream.

Izorkin commented:
@mjankowski re-open PR so i can update it?

Izorkin commented:
@mjankowski update.

mjankowski commented:
Thanks for update - sorry about any confusion earlier.

Can you update the PR title and description to reflect whatever the current state of changes is? (I'm not an nginx expert - just trying to move some PRs along).

Izorkin commented:
@mjankowski updated description and title.

mjankowski commented:
> @mjankowski updated description and title.

Thanks -- not sure we can get reviewed/merge, but this strikes me as at least much more focused/reviewable at this point.

ClearlyClaire commented on dist/nginx.conf:
```diff
@@ -65,60 +65,55 @@ server {
-  # then needed must replace line `try_files $uri =404;` with `try_files $uri @proxy;`.
-  location = /sw.js {
-    add_header Cache-Control "public, max-age=604800, must-revalidate";
     add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
```
afaik, Rails already sets HSTS headers… wouldn't this be an issue?

Izorkin replied:
In this case, these headers do not affect the `@mastodon` block

Izorkin commented:
Rebased PR.

Izorkin committed:
nginx: remove try_files directive

There is no need to use the try_files directive in every location. In addition,
it causes unnecessary system calls.

Izorkin commented:
cc @ClearlyClaire @renchap @Gargron

Izorkin commented:
Thanls!
